src/optimize.rs

```diff
@@ -101,52 +125,134 @@ fn get_self_cost(egraph: &EGraph<Mdl, TensorAnalysis>, enode: &Mdl) -> f32 {
+                let t_b = _b_data.meta;
+
+                // Get op
+                let op = (*g.model).get_or_create_element(OpType_OP_EW_MUL, t_a, t_b);
+                assert!(op != Op_INVALID_OP);
+                (*op.ptr).runtime.clone()
+            }
         }
-    }
-}
 
-/*
-struct Cost;
-impl CostFunction<Mdl> for Cost {
-    type Cost = (f64, Vec<usize>);
-    fn cost<C: FnMut(Id) -> Self::Cost>(&mut self, enode: &Mdl, mut costs: C) -> Self::Cost {
-        let children_sizes = enode.fold(vec![], |mut sizes, id| {
-            sizes.push(costs(id).1);
-            sizes
-        });
-        layouts(enode)
-            .into_iter()
-            .map(|layout| Self::run_time(enode, layout, &children_sizes))
-            .min_by(|(x, _), (y, _)| x.partial_cmp(y).unwrap())
-            .unwrap()
-        // TODO gotta calc output sizes
-    }
-}
+        Mdl::Matmul([_act, _a, _b]) => {
+            // Check types
+            let _act_data = x(_act);
+            let _a_data = x(_a);
+            let _b_data = x(_b);
+            assert!(_act_data.dtype == DataKind::Scalar);
+            assert!(_a_data.dtype == DataKind::Tnsr);
+            assert!(_b_data.dtype == DataKind::Tnsr);
 
-struct Layout;
-fn layouts(_e: &Mdl) -> Vec<Layout> {
-    todo!()
-}
+            unsafe {
+                // Get arguments
+                let t_a = *_a_data.meta;
+                let t_b = *_b_data.meta;
+                let activation: ActiMode = _act_data.val.try_into().unwrap();
+
+                // Get op
+                let op = (*g.model).get_or_create_matmul(t_a, t_b, activation);
+                assert!(op != Op_INVALID_OP);
+                (*op.ptr).runtime.clone()
+            }
+        }
 
-impl Cost {
-    fn run_time(
-        _e: &Mdl,
-        _layout: Layout,
-        _sizes: &[Vec<usize>],
-    ) -> <Cost as CostFunction<Mdl>>::Cost {
-        todo!()
+        Mdl::Concat([_axis, _ndim, _a, _b]) => {
+            // Check types
+            let _axis_data = x(_axis);
+            let _ndim_data = x(_ndim);
+            let _a_data = x(_a);
+            let _b_data = x(_b);
+            assert!(_axis_data.dtype == DataKind::Scalar);
+            assert!(_ndim_data.dtype == DataKind::Scalar);
+            assert!(_a_data.dtype == DataKind::Tnsr);
+            assert!(_b_data.dtype == DataKind::Tnsr);
+
+            unsafe {
+                // Get arguments
+                let t_a = *_a_data.meta;
+                let t_b = *_b_data.meta;
+                let axis = _axis_data.val;
+                let ndim = _ndim_data.val;
+
+                // Pass ownership to C++
+                let mut inputs = vec![t_a, t_b];
+                inputs.shrink_to_fit();
+                assert!(inputs.len() == inputs.capacity());
+                let ptr = inputs.as_mut_ptr();
+                std::mem::forget(inputs);
+
+                // Get op
+                let need_copy = vec![false, false].as_mut_ptr();
```

Comment:
The vector might be de-allocated before you use the pointer to its data. You need to assign it to a variable.

Also, you can use a slice instead of a vec:
```
let mut need_copy = [false, false];
let op = ...(..., need_copy.as_mut_ptr());
```

Reply:
Done